src/libdrakvuf/win-processes.c

```diff
@@ -0,0 +1,148 @@
+ * Covered Software stating terms other than these, you may choose to use  *
+ * and redistribute Covered Software under those terms instead of these.   *
+ *                                                                         *
+ * Source is provided to this software because we believe users have a     *
+ * right to know exactly what a program is going to do before they run it. *
+ * This also allows you to audit the software for security holes.          *
+ *                                                                         *
+ * Source code also allows you to port DRAKVUF to new platforms, fix bugs, *
+ * and add new features.  You are highly encouraged to submit your changes *
+ * on https://github.com/tklengyel/drakvuf, or by other methods.           *
+ * By sending these changes, it is understood (unless you specify          *
+ * otherwise) that you are offering unlimited, non-exclusive right to      *
+ * reuse, modify, and relicense the code.  DRAKVUF will always be          *
+ * available Open Source, but this is important because the inability to   *
+ * relicense code has caused devastating problems for other Free Software  *
+ * projects (such as KDE and NASM).                                        *
+ * To specify special license conditions of your contributions, just say   *
+ * so when you send them.                                                  *
+ *                                                                         *
+ * This program is distributed in the hope that it will be useful, but     *
+ * WITHOUT ANY WARRANTY; without even the implied warranty of              *
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the DRAKVUF   *
+ * license file for more details (it's in a COPYING file included with     *
+ * DRAKVUF, and also available from                                        *
+ * https://github.com/tklengyel/drakvuf/COPYING)                           *
+ *                                                                         *
+ ***************************************************************************/
+
+#include <libvmi/libvmi.h>
+#include <stdlib.h>
+#include <string.h>
+#include <errno.h>
+#include <sys/mman.h>
+#include <stdio.h>
+#include <glib.h>
+
+#include "vmi.h"
+
+addr_t drakvuf_get_current_thread(drakvuf_t drakvuf, uint64_t vcpu_id){
+    vmi_instance_t vmi = drakvuf->vmi;
+    addr_t thread;
+
+    size_t prcb;
+
+    /*
+     * fs_base/gs_base in the info->regs structure are not actually filled in
+     * by Xen for vm_events, so we need to manually ask for these each time
+     */
+    reg_t fsgs;
+    if(vmi_get_page_mode(vmi) == VMI_PM_IA32E)  {
+        vmi_get_vcpureg(vmi, &fsgs, GS_BASE, vcpu_id);
+        prcb=offsets[KPCR_PRCB];
+    } else {
+        vmi_get_vcpureg(vmi, &fsgs, FS_BASE, vcpu_id);
+        prcb=offsets[KPCR_PRCBDATA];
+    }
+
+    vmi_read_addr_va(vmi, fsgs + prcb + offsets[KPRCB_CURRENTTHREAD],
```

Comment:
We should check the return of this function if it's VMI_SUCCESS and set thread explicitly to 0 if it is not. Or initializing thread to 0 would also get the same effect, though explicitly checking is a bit cleaner.